subsys/llext/llext.c

```diff
@@ -324,3 +324,189 @@ void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_da
+	struct llext *next, *tmp, *first = ext[0], *last = ext[n_ext - 1];
+	unsigned int i, j, n_map, n_exp_tab;
+	int ret;
+
+	if (!map) {
+		LOG_ERR("cannot allocate list of maps of %zu", sizeof(*map) * n_ext);
+		return -ENOMEM;
+	}
+
+	/*
+	 * Each extension has a section map, so if this loop completes
+	 * successfully in the end n_map == n_ext. But if it's terminated
+	 * because of an allocation failure, we need to know how many maps have
+	 * to be freed.
+	 */
+	for (n_map = 0, n_exp_tab = 0; n_map < n_ext; n_map++) {
+		/* Need to allocate individually, because that's how they're freed */
+		map[n_map] = llext_alloc(sizeof(**map) * ext[n_map]->sect_cnt);
+		if (!map[n_map]) {
+			LOG_ERR("cannot allocate section map of %zu",
+				sizeof(**map) * ext[n_map]->sect_cnt);
+			ret = -ENOMEM;
+			goto free_map;
+		}
+
+		/* Not every extension exports symbols, count those that do */
+		if (ext[n_map]->exp_tab.sym_cnt) {
+			n_exp_tab++;
+		}
+	}
+
+	/* Array of pointers to exported symbol tables. Can be NULL if n_exp_tab == 0 */
+	struct llext_symbol **exp_tab = llext_alloc(sizeof(*exp_tab) * n_exp_tab);
+
+	if (n_exp_tab) {
+		if (!exp_tab) {
+			LOG_ERR("cannot allocate list of exported symbol tables of %zu",
+				sizeof(*exp_tab) * n_exp_tab);
+			ret = -ENOMEM;
+			goto free_map;
+		}
+
+		/* Now actually allocate new exported symbol lists */
+		for (i = 0, j = 0; i < n_ext; i++) {
+			if (ext[i]->exp_tab.sym_cnt) {
+				size_t size = sizeof(**exp_tab) * ext[i]->exp_tab.sym_cnt;
+
+				exp_tab[j] = llext_alloc(size);
+				if (!exp_tab[j]) {
+					LOG_ERR("cannot allocate exported symbol table of %zu",
+						size);
+					ret = -ENOMEM;
+					goto free_sym;
+				}
+				memcpy(exp_tab[j++], ext[i]->exp_tab.syms, size);
+			}
+		}
+	}
+
+	k_mutex_lock(&llext_lock, K_FOREVER);
+
+	/* Allocate extensions and add them to the global list */
+	for (i = 0, j = 0; i < n_ext; i++) {
+		next = llext_alloc(sizeof(*next));
+		if (!next) {
+			LOG_ERR("cannot allocate LLEXT of %zu", sizeof(*next));
+			ret = -ENOMEM;
+			goto free_llext;
+		}
+
+		*next = *ext[i];
+		ext[i] = next;
+		if (next->exp_tab.sym_cnt) {
+			next->exp_tab.syms = exp_tab[j++];
+		}
+
+		sys_slist_append(&_llext_list, &next->_llext_list);
+	}
+
+	k_mutex_unlock(&llext_lock);
+
+	/* Copy section maps */
+	for (i = 0; i < n_ext; i++) {
+		size_t map_size = sizeof(struct llext_elf_sect_map) * ext[i]->sect_cnt;
+
+		memcpy(map[i], ldr[i]->sect_map, map_size);
+		ldr[i]->sect_map = map[i];
+	}
+
+	llext_free(map);
+	llext_free(exp_tab);
+
+	/* Restore dependencies */
+	SYS_SLIST_FOR_EACH_CONTAINER(&_llext_list, next, _llext_list) {
+		for (j = 0; next->dependency[j] && j < ARRAY_SIZE(next->dependency); j++) {
+			if (next->dependency[j] < first || next->dependency[j] >= last) {
+				/* Inconsistency detected */
+				LOG_ERR("dependency out of range");
+				ret = -EINVAL;
+				goto free_locked;
+			}
+
+			next->dependency[j] = llext_by_name(next->dependency[j]->name);
```

Comment:
This seems fairly costly, we're doing two loops and then walking the list of extensions comparing strings along the way. Had we used array indices instead here maybe we could have avoided all this work?

Reply:
@teburd well, the loops are quite short. The most outer and the most inner loops are over all extensions, of which some use-cases have just 1, in SOF case we can have 5 or possibly up to 10. The middle loop over dependencies so far hardly executes more than once, not in SOF in any case, and I'm not sure if anybody else uses dependencies. So, it isn't that expensive. Not sure we need to switch to an array just because of this.